# MirrorPrimerChecker

This script checks the suitability of primers using the **MIrROR v.2.0** database.
<br><br><br>

## üìå Features
- Accepts primer sequences (IUPAC codes allowed) in the 5'‚Üí3' direction for both forward and reverse primers.
- Matches input primers against sequences in the database.
- Outputs matched sequences and their metadata.

<br><br>
## üì• Installation
Make sure you have Python installed, then install required dependencies:

```sh
pip install pandas
```

<br><br>
## üöÄ Usage
### Run the script with:

```sh
python main.py <forward_primer> <reverse_primer> --db_path <database.tsv> --output_dir <output_directory>
```

### Example:
```sh
python main.py CAGCMGCCGCGGTAA CCRAMCTGTCTCACGACG --db_path database.tsv --output_dir results
```
cf. Example primer set: A519F (5‚Äô‚ÄêCAGCMGCCGCGGTAA‚Äê3‚Ä≤), U2428R (5‚Äô-CCRAMCTGTCTCACGACG-3‚Ä≤)
<br><br>
### Arguments
| Argument           | Description |
|--------------------|-------------|
| `forward_primer`  | Forward primer sequence (5'‚Üí3') |
| `reverse_primer`  | Reverse primer sequence (5'‚Üí3') |
| `--db_path`       | Path to the MIRROR primer database (TXT format) |
| `--output_dir`    | Directory to save output files (default: `output/`) |

<br><br>
## üìÇ Output Files
| File                          | Description |
|--------------------------------|-------------|
| `1_Summary.txt`               | Summary of results, including Domain Distribution, match percentage of Operon Sequences, number of matched sequences, total database coverage, number of matched genomes and total genomes in the database, and number of matched species and total species in the database |
| `2_Query_primer_sequences.txt`  | Lists all possible sequences generated by considering IUPAC codes |
| `3_Database_matches.tsv`        | Lists matched sequences along with taxonomic information and the in silico amplicon sequences with their lengths |


<br><br>
## üîç Database Download
You can download the original MIrROR v.2.0 database and Minimap2 index (.mmi) files from [Zenodo](https://doi.org/10.5281/zenodo.14603319).

<br><br>
## üìñ Example Output

### 1_Summary.txt
```
# Input Primers:
	- Forward Primer: 5'‚Üí3' CAGCAGCCGCGGTAA
	- Reverse Primer: 5'‚Üí3' CGTCGTGAGACAGGTCGG

# Domain Distribution:
	- Bacteria: 463520 (99.56%)
	- Archaea: 2040 (0.44%)

# Operon Sequences:
	- Percentage of Matches: 97.69%
	- Matched: 465560
	- Total Covered: 476579

# Genomes:
	- Matched: 242270
	- Total Covered: 249907

# Species:
	- Matched: 25479
	- Total Covered: 29051
```

### 2_Query_primer_sequences.txt
```
# Forward Primer Sequence(s)
CAGCAGCCGCGGTAA
CAGCCGCCGCGGTAA

# Reverse Primer Sequence(s)
CGTCGTGAGACAGGTCGG
CGTCGTGAGACAGGTTGG
CGTCGTGAGACAGTTCGG
CGTCGTGAGACAGTTTGG
```

### 3_Database_matches.tsv

| Accession        | GenBankAccession | OperonNumber | Domain   | Phylum    | Class   | Order            | Family           | Genus          | Species                   | Amplicon_Sequence | Amplicon_Length | Forward_Primer | Reverse_Primer |
|-----------------|----------------|-------------|---------|----------|--------|----------------|----------------|---------------|--------------------------|------------------|---------------|---------------|---------------|
| GCA_014495985.1 | CP061133.1      | operon_1    | Bacteria | Bacillota | Bacilli | Lactobacillales | Streptococcaceae | Streptococcus | Streptococcus pyogenes   | CAGCAGCCGC ... GACAGTTCGG     | 4012          | CAGCAGCCGCGGTAA | CGTCGTGAGACAGTTCGG |

<br><br>
## üìö References
Martijn, Joran, et al. "Confident phylogenetic identification of uncultured prokaryotes through long read amplicon sequencing of the 16S‚ÄêITS‚Äê23S rRNA operon." Environmental microbiology 21.7 (2019): 2485-2498.

<br><br>
## üì¢ License
- Research and non-commercial use allowed.
- When using in conjunction with the MIrROR database, please include the following citation:
```
Lee, J., Cho, S., & Kim, H. MIrROR (2.0) [Data set]. (2025). https://doi.org/10.5281/zenodo.14603319
```
